types/proposal_test.go

```diff
@@ -191,3 +192,65 @@ func TestProposalProtoBuf(t *testing.T) {
 		}
 	}
 }
+
+func TestIsTimely(t *testing.T) {
+	genesisTime, err := time.Parse(time.RFC3339, "2019-03-13T23:00:00Z")
+	require.NoError(t, err)
+	testCases := []struct {
+		name         string
+		proposalTime time.Time
+		localTime    time.Time
+		precision    time.Duration
+		msgDelay     time.Duration
+		expectTimely bool
+	}{
+		{
+			// Checking that the following inequality evaluates to true:
+			// 1 - 2 < 0 < 1 + 2 + 1
+			name:         "basic timely",
```

Comment:
For me, the "basic" case, meaning the most common, is:
 * The proposal timestamp is "for sure" in the past, so: proposalTime + Precision < localTime - Precision
 * The receiving time of the proposal is "fore sure" within the acceptable window, so: localTime + Precision < proposalTime + msgDelay - Precision